sw/device/lib/crypto/aes_gcm/aes_gcm.c

```diff
@@ -0,0 +1,351 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/crypto/drivers/aes.h"
+
+enum {
+  /* Log2 of the number of bytes in an AES block. */
+  kAesBlockLog2NumBytes = 4,
+};
+OT_ASSERT_ENUM_VALUE(kAesBlockNumBytes, 1 << kAesBlockLog2NumBytes);
+
+/**
+ * Precomputed modular reduction constants for Galois field multiplication.
+ *
+ * This implementation uses 4-bit windows. The bytes here represent 12-bit
+ * little-endian values.
+ *
+ * The entry with index i in this table is equal to i * 0xe1, where the bytes i
+ * and 0xe1 are interpreted as polynomials in the GCM Galois field. For
+ * example, 0xe1 = 11100001 is the degree-8 polynomial x^8 + x^2 + x + 1.
+ *
+ * The field modulus for GCM is 2^128 + x^8 + x^2 + x + 1. Therefore, if a
+ * field element is shifted, it can be reduced by multiplying the high bits
+ * (128 and above) by 0xe1 and adding them to the low bits.
+ *
+ * There is a size/speed tradeoff in window size. For 8-bit windows, GHASH
+ * becomes significantly faster, but the overhead for computing the product
+ * table of a given hash subkey becomes higher, so larger windows are slower
+ * for smaller inputs but faster for large inputs.
+ */
+static const uint16_t kGFReduceTable[16] = {
+    0x0000, 0x201c, 0x4038, 0x6024, 0x8070, 0xa06c, 0xc048, 0xe054,
+    0x00e1, 0x20fd, 0x40d9, 0x60c5, 0x8091, 0xa08d, 0xc0a9, 0xe0b5};
+
+/**
+ * Performs a bitwise XOR of two blocks.
+ *
+ * This operation corresponds to addition in the Galois field.
+ *
+ * @param condition Either 0 or 1
+ * @param x First operand block
+ * @param y Second operand block
+ * @param out Buffer in which to store output; can be the same as one or both
+ * operands.
+ */
+static inline void block_xor(const aes_block_t *x, const aes_block_t *y,
+                             aes_block_t *out) {
+  for (size_t i = 0; i < kAesBlockNumWords; i++) {
+    out->data[i] = x->data[i] ^ y->data[i];
+  }
+}
+
+/**
+ * Reverse the bits of a 4-bit number.
+ *
+ * @param word Input word.
+ * @return Word with bytes reversed.
+ */
+static uint8_t reverse_bits(uint8_t byte) {
+  uint8_t out = 0;
+  for (size_t i = 0; i < 4; i++) {
+    out <<= 1;
+    out |= (byte >> i) & 1;
+  }
+  return out;
+}
```

Comment:
How about something like:
```
byte = (byte & 0x55) << 1 | (byte & 0xaa) >> 1;
byte = (byte & 0x33) << 2 | (byte & 0xcc) >> 2;
byte = (byte & 0x0f) << 4 | (byte & 0xf0) >> 4;
return byte;
```
This is the byte version of the first algorithm in Hacker's Delight 7.1, 2nd ed ([link](https://learning.oreilly.com/library/view/hackers-delight-second/9780133084993/ch07.html#:-:text=x%20%3D%20(x%20%26%200x55555555)%C2%A0%C2%A0%3C%3C%C2%A0%C2%A0%C2%A01,16%20%7C%20(x%20%26%200xFFFF0000)%20%3E%3E%2016%3B), also some discussion on [SO](https://stackoverflow.com/a/2602885)).

Side note: we should also consider the bitmanip extensions.